pkg/controller/master/upgradelog/controller.go

```
@@ -771,24 +773,48 @@ func (h *handler) stopCollect(upgradeLog *harvesterv1.UpgradeLog) error {
 func (h *handler) cleanup(upgradeLog *harvesterv1.UpgradeLog) error {
 	// Cleanup the relationship from its corresponding Upgrade resource
 	upgradeName := upgradeLog.Spec.UpgradeName
-	upgrade, err := h.upgradeCache.Get(util.HarvesterSystemNamespaceName, upgradeName)
-	if err != nil {
-		if apierrors.IsNotFound(err) {
-			return nil
+	var err1, err2, err3 error
+	errorStr := fmt.Sprintf("upgradeLog %s/%s upgrade %s cleanup errors", upgradeLog.Namespace, upgradeLog.Name, upgradeName)
+
+	upgrade, err1 := h.upgradeCache.Get(util.HarvesterSystemNamespaceName, upgradeName)
+	for i := 0; i < 1; i++ {
+		if err1 != nil {
+			// can't return directly, need to clean other related resources
+			if apierrors.IsNotFound(err1) {
+				err1 = nil
+				break
+			}
+			errorStr = fmt.Sprintf("%s failed to get upgrade %s", errorStr, err1.Error())
+			break
+		}
+		// already updated
+		if upgrade.Status.UpgradeLog == "" {
+			break
+		}
+		upgradeToUpdate := upgrade.DeepCopy()
+		upgradeToUpdate.Status.UpgradeLog = ""
+		_, err2 = h.upgradeClient.Update(upgradeToUpdate)
+		if err2 != nil {
+			if apierrors.IsNotFound(err2) {
+				err2 = nil
+				break
+			}
+			errorStr = fmt.Sprintf("%s failed to update upgrade %s", errorStr, err2.Error())
+			break
 		}
-		return err
 	}
-	upgradeToUpdate := upgrade.DeepCopy()
-	upgradeToUpdate.Status.UpgradeLog = ""
-	if _, err = h.upgradeClient.Update(upgradeToUpdate); err != nil {
-		return err
+
+	// Remove all if the UpgradeLog resource is deleted before normal tear down
+	logrus.Info("Removing all other related resources")
+	err3 = h.stopCollect(upgradeLog)
+	if err3 != nil {
+		errorStr = fmt.Sprintf("%s failed to clean other resources %s", errorStr, err3.Error())
```

Consider wrapping the errors with `fmt.Errorf("%w", err)` rather than accumulating error strings that can not be checked for individual errors later.

---

This clean up function is special, it can't return on error simply, but need to try releasing resource as more as possible.

That's why we try to put those errors together.

---

Yes, I understand, but the individual errors are swallowed this way.
By using `fmt.Sprintf("... %w", err)` you can create an error that still matches on the original error (e.g. when you later to a `errors.IsNotFound(err)` from the caller of this function).

---

Good idea, updated.